File: examples/python/customer/handlers/add_loyalty_points.py
"""Handler for AddLoyaltyPoints command."""

from datetime import datetime, timezone

import structlog
from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

from angzarr import angzarr_pb2 as angzarr
from proto import domains_pb2 as domains

from handlers import CommandRejectedError


def handle_add_loyalty_points(
    command_book: angzarr.CommandBook,
    command_any: Any,
    state: domains.CustomerState,
    seq: int,
    log: structlog.BoundLogger,
) -> angzarr.EventBook:
    """Handle AddLoyaltyPoints command."""
    if not state.name:
        raise CommandRejectedError("Customer does not exist")

    cmd = domains.AddLoyaltyPoints()
    command_any.Unpack(cmd)

    if cmd.points <= 0:
        raise CommandRejectedError("Points must be positive")

    new_balance = state.loyalty_points + cmd.points

    log.info(
        "adding_loyalty_points",
        points=cmd.points,
        new_balance=new_balance,
        reason=cmd.reason,
    )

    event = domains.LoyaltyPointsAdded(
        points=cmd.points,
        new_balance=new_balance,
        reason=cmd.reason,
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[
            angzarr.EventPage(
                num=seq,
                event=event_any,
                created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
            )
        ],
    )

File: examples/rust/customer/src/handlers/add_loyalty_points.rs
//! Handler for AddLoyaltyPoints command.

use angzarr::proto::{CommandBook, EventBook};
use common::proto::{AddLoyaltyPoints, CustomerState, LoyaltyPointsAdded};
use common::{decode_command, require_exists, require_positive, Result};

use crate::errmsg;
use crate::state::build_event_response;

/// Handle the AddLoyaltyPoints command.
///
/// Adds loyalty points to an existing customer's balance.
/// Fails if the customer does not exist or points are not positive.
pub fn handle_add_loyalty_points(
    command_book: &CommandBook,
    command_data: &[u8],
    state: &CustomerState,
    next_seq: u32,
) -> Result<EventBook> {
    require_exists(&state.name, errmsg::CUSTOMER_NOT_FOUND)?;

    let cmd: AddLoyaltyPoints = decode_command(command_data)?;

    require_positive(cmd.points, errmsg::POINTS_POSITIVE)?;

    let new_balance = state.loyalty_points + cmd.points;
    let new_lifetime_points = state.lifetime_points + cmd.points;

    let event = LoyaltyPointsAdded {
        points: cmd.points,
        new_balance,
        reason: cmd.reason,
        new_lifetime_points, // Fact: total lifetime points after this event
    };

    Ok(build_event_response(
        state,
        command_book.cover.clone(),
        next_seq,
        "type.examples/examples.LoyaltyPointsAdded",
        event,
    ))
}
